### PR TITLE
OTT-908: Add support to get country from IP address 

### PR DESCRIPTION
**Pull request by Ankit-Pinge**

### CODE REVIEW

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 1]
```diff
@@ -25,7 +27,7 @@ func getFilteredBidders(rCtx models.RequestCtx, bidRequest *openrtb2.BidRequest,
 	if !ok {
 		return filteredBidders, false
 	}
-	data := generateEvaluationData(bidRequest)
+	data := generateEvaluationData(bidRequest, rCtx)
```
Comment (ashishshinde-pubm) [inline modules/pubmatic/openwrap/traffic_shapping.go:30]: generateEvaluationData(bidRequest, rCtx)
==> generateEvaluationData( rCtx,bidRequest) 
pass it as first argument

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 2]
```diff
@@ -47,22 +49,36 @@ func getFilteredBidders(rCtx models.RequestCtx, bidRequest *openrtb2.BidRequest,
 	return filteredBidders, allPartnersFilteredFlag
 }
 
-func generateEvaluationData(BidRequest *openrtb2.BidRequest) map[string]interface{} {
+func generateEvaluationData(BidRequest *openrtb2.BidRequest, rCtx models.RequestCtx) map[string]interface{} {
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/traffic_shapping.go:52]: If you are passing rCtx models.RequestCtx here, no need to pass BidRequest, its part of rCtx.

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 3]
```diff
@@ -47,22 +49,36 @@ func getFilteredBidders(rCtx models.RequestCtx, bidRequest *openrtb2.BidRequest,
...
+	}
+	if ip == "" && bidRequest.Device != nil && bidRequest.Device.IP != "" {
+		ip = bidRequest.Device.IP
+	}
+	if ip == "" && bidRequest.Device != nil && bidRequest.Device.IPv6 != "" {
+		ip = bidRequest.Device.IPv6
+	}
+
+	if ip != "" {
+		if country, err := getCountryFromIP(ip, rCtx.GeoLooker); err == nil {
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/traffic_shapping.go:78]: Can we log this error?

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 4]
```diff
@@ -47,22 +49,36 @@ func getFilteredBidders(rCtx models.RequestCtx, bidRequest *openrtb2.BidRequest,
...
 	}
-
 	if bidRequest.User != nil && bidRequest.User.Geo != nil && bidRequest.User.Geo.Country != "" {
 		return bidRequest.User.Geo.Country
 	}
 
-	// return country using  netacuity
+	ip := ""
+	if rCtx.IP != "" {
+		ip = rCtx.IP
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/traffic_shapping.go:68]: Are we setting this rCtx.IP somewhere?

**Review by ashishshinde-pubm — APPROVED** [review]
LGTM

File: modules/pubmatic/openwrap/traffic_shapping_test.go [hunk 5]
```diff
@@ -429,3 +432,175 @@ func TestGetFilteredBidders(t *testing.T) {
...
+		geoLooker geodb.Geography
+	}
+	tests := []struct {
+		name  string
+		args  args
+		setup func()
+		want  string
+		err   error
+	}{
+		{
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/traffic_shapping_test.go:572]: Can you add invalid ip scenario?

File: modules/pubmatic/openwrap/traffic_shapping.go (hunk 1 above)
Comment (Ankit-Pinge) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Done

File: modules/pubmatic/openwrap/traffic_shapping.go (hunk 3 above)
Comment (Ankit-Pinge) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Added

File: modules/pubmatic/openwrap/traffic_shapping.go (hunk 4 above)
Comment (Ankit-Pinge) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Yes we we setting this in entrypointhook.
The IP address priority is as follows:

1. `HTTP_RLNCLIENTIPADDR  //For API
2. 	HTTP_X_FORWARDED_IP
3. 	HTTP_X_CLUSTER_CLIENT_IP
4. 	HTTP_SOURCE_IP
5. 	REMOTE_ADDR`

File: modules/pubmatic/openwrap/traffic_shapping_test.go (hunk 5 above)
Comment (Ankit-Pinge) [reply modules/pubmatic/openwrap/traffic_shapping_test.go:572]: Added

**Review by Pubmatic-Dhruv-Sonone — APPROVED** [review]

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 6]
```diff
@@ -74,3 +90,14 @@ func evaluateBiddingCondition(data, rules interface{}) bool {
 	}
 	return output == true
 }
+
+func getCountryFromIP(ip string, geoInfoFetcher geodb.Geography) (string, error) {
+	if geoInfoFetcher == nil {
+		return "", errors.New("geoDB instance is missing")
```
Comment (pm-viral-vala) [inline modules/pubmatic/openwrap/traffic_shapping.go:96]: can we create an `error variable` instead of creating `errors.New` everytime.

**Review by pm-viral-vala — APPROVED** [review]

**Review by ashishshinde-pubm — APPROVED** [review]

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 7]
```diff
@@ -74,3 +90,14 @@ func evaluateBiddingCondition(data, rules interface{}) bool {
 	}
 	return output == true
 }
+
+func getCountryFromIP(ip string, geoInfoFetcher geodb.Geography) (string, error) {
```
[the remark this answers is not in the document]
Comment (Ankit-Pinge) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Done

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 8]
```diff
@@ -47,22 +49,36 @@ func getFilteredBidders(rCtx models.RequestCtx, bidRequest *openrtb2.BidRequest,
...
+		ip = bidRequest.Device.IP
+		if ip == "" {
+			ip = bidRequest.Device.IPv6
+		}
+	}
+
+	if ip != "" {
+		country, err := getCountryFromIP(ip, rCtx.GeoInfoFetcher)
+		if err != nil {
+			glog.Errorf("Error while fetching country from IP [%s] | Error[%s]", ip, err)
```
[the remark this answers is not in the document]
Comment (Ankit-Pinge) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Done

**Pull request merged**